File: exam/src/strategies/validation/DocumentValidator.ts
import { Document } from "../../objects/Document";

export abstract class DocumentValidator {
    protected readonly DOCUMENT_MIN_SIZE = 0;
    protected readonly DOCUMENT_MIN_NAME_LENGTH = 0;

    public validateSize(size: number): boolean {
        return size > this.DOCUMENT_MIN_SIZE;
    }

    public validateExtension(extension: string): boolean {
        return extension.length > this.DOCUMENT_MIN_NAME_LENGTH && extension.charAt(0) == '.';
    }

    public validateFileName(filename: string): boolean {
        return filename.length > this.DOCUMENT_MIN_NAME_LENGTH;
    }

    public abstract validateMetadata(metadata: Record<string,string>): boolean;
}

File: exam/src/validator/DocumentValidator.java
package validator;

import document.Document;

import java.util.ArrayList;
import java.util.List;

public abstract class DocumentValidator {

    public DocumentValidator() {
    }

    public ValidationResult validate(Document document) {
        boolean valid = true;
        List<String> errors = new ArrayList<>();

        if (!validateBasicRequirements(document)) {
            valid = false;
            errors.add("Basic requirement not met: missing filename");
        }
        if (!validateSpecificRequirements(document, errors)) {
            valid = false;
        }

        return new ValidationResult(valid, errors);
    }

    private boolean validateBasicRequirements(Document document) {
        return document.hasFilename();
    }

    private boolean validateSpecificRequirements(Document document, List<String> errors) {
        if (!validateSize(document)) {
            errors.add("Specific requirement not met: invalid size");
            return false;
        }
        if (!validateExtension(document)) {
            errors.add("Specific requirement not met: invalid extension");
            return false;
        }
        if (!validateMetadata(document)) {
            errors.add("Specific requirement not met: invalid metadata");
            return false;
        }
        return true;
    }

    protected abstract boolean validateSize(Document document);

    protected abstract boolean validateExtension(Document document);

    protected abstract boolean validateMetadata(Document document);
}
